### PR TITLE
Hashed Time-Locked Contracts

### PR DESCRIPTION
**Pull request by jmjatlanta**
Changes to implement [BSIP 44](https://github.com/bitshares/bsips/blob/master/bsip-0044.md)

Fixes issue #1468

### CODE REVIEW

File: libraries/chain/htlc_evaluator.cpp [hunk 1]
```diff
@@ -31,15 +32,22 @@ namespace graphene {
...
 
-          FC_ASSERT( fc::time_point_sec(o.epoch) > db().head_block_time() );
-          // make sure we have the funds for the HTLC
-          FC_ASSERT( db().get_balance( o.source, o.amount.asset_id ) >= (o.amount ) );
-          // make sure we have the funds for the fee
-          FC_ASSERT( db().get_balance( o.source, asset().asset_id) > o.fee );
-          return void_result();
+         const graphene::chain::global_property_object& global_properties = db().get_global_properties();
+         // make sure the expiration is reasonable
+         FC_ASSERT( o.epoch.sec_since_epoch() < fc::time_point::now().sec_since_epoch()
```
Comment (pmconrad) [inline libraries/chain/htlc_evaluator.cpp]: Should use ``head_block_time()`` instead of ``now()``

File: libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp [hunk 2]
```diff
@@ -63,14 +62,47 @@ namespace graphene { namespace chain {
       uint16_t                accounts_per_fee_scale              = GRAPHENE_DEFAULT_ACCOUNTS_PER_FEE_SCALE; ///< number of accounts between fee scalings
       uint8_t                 account_fee_scale_bitshifts         = GRAPHENE_DEFAULT_ACCOUNT_FEE_SCALE_BITSHIFTS; ///< number of times to left bitshift account registration fee at each scaling
       uint8_t                 max_authority_depth                 = GRAPHENE_MAX_SIG_CHECK_DEPTH;
+
+      struct ext
+      {
+         struct committee_updatable_parameters
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: Also need hardfork protection in ``committee_member_update_global_parameters_evaluator``, these extensions must remain unset until hardfork.

[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: Also need hardfork protection in `proposal_create_evaluator`.

File: libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp [hunk 3]
```diff
@@ -30,6 +30,12 @@ namespace graphene { namespace chain { struct fee_schedule; } }
 
 namespace graphene { namespace chain {
 
+   struct committee_updatable_parameters
```
Comment (abitmore) [inline libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: If name it `committee_updatable_parameters` then it itself should be extendable, but currently it's not. (btw looks like the name is inspired by YOYOW)

An approach is to entirely replace `ext` with it and put optional fields inside, but then the struct with all fields are optional is not good for calls like `get_global_properties`.

So perhaps better to use a struct for objects to be stored in object database, and use another extendable struct for operations, and map the fields. ([YOYOW's approach](https://github.com/yoyow-org/yoyow-core/blob/v0.3.0-180626/libraries/chain/include/graphene/chain/protocol/committee_member.hpp#L164))

Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: @abitmore  These questions/comments will expose my ignorance. Sorry for the required mentoring, but here goes:

Your first paragraph seems to say that my naming is inconsistent, and inconsistent with Bitshares.  So if I change the name, does that negate your next two paragraphs?

Your second paragraph says (to me): "make the committee_updatable_parameters struct extendable. But that will make get_global_properties not work or be messy".

The proposed solution in paragraph 3 says (to me): "Look at how YOYOW did it. They use structs for persistence, and extendables for operations, and then map back and forth within operations."

Am I interpreting correctly?

Comment (abitmore) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: Since there is an `extension` field in the object, `get_global_properties` is still a bit messy. I guess I can live with that, although it's best if we have https://github.com/bitshares/bitshares-core/issues/167 done.

Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: Agreed that https://github.com/bitshares/bitshares-core/issues/167 would be nice. Perhaps I'll tackle that next.

File: libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: I am not averse to adding hardfork protection in proposal_create_evaluator, but I have a question.

Currently the call will throw, as the committee parameters are not set, and can not be set until after hardfork time. That is a round-about way to handle hardfork protection, but it is less code.

So which is better: explicit hardfork protection with more code, or implicit with less code?

[the remark this answers is not in the document]
Comment (cogutvalera) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: less code - less bugs :)

[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: @jmjatlanta we need (duplicate) hardfork protection code in `proposal_create_evaluator.do_evaluate` because `committee_member_update_global_parameters_evaluator.do_evaluate` won't be called when there is a `committee_member_update_global_parameters_operation` in a proposal.

[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/chain_parameters.hpp]: Done. See https://github.com/bitshares/bitshares-core/pull/1370/commits/e8faf8115455d451c843514cec4137c44de07f75

File: libraries/wallet/wallet.cpp [hunk 4]
```diff
@@ -3012,6 +3133,67 @@ uint64_t wallet_api::get_asset_count()const
    return my->_remote_db->get_asset_count();
 }
 
+std::vector<unsigned char> string_to_vec(std::string incoming)
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: fc already has that: https://github.com/bitshares/bitshares-fc/blob/master/src/variant.cpp#L664

File: libraries/wallet/wallet.cpp [hunk 5]
```diff
@@ -3012,6 +3133,67 @@ uint64_t wallet_api::get_asset_count()const
...
+
+signed_transaction wallet_api::htlc_prepare( string source, string destination, string asset_symbol, string amount,
+         string hash_algorithm, const std::string& preimage_hash, size_t preimage_size, 
+         const uint32_t seconds_in_force, bool broadcast)
+{
+   return my->htlc_prepare(source, destination, asset_symbol, amount, hash_algorithm, string_to_vec(preimage_hash), preimage_size,
+         seconds_in_force, broadcast);
+}
+
+std::string object_id_to_string(const graphene::db::object_id_type& obj_id)
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: Can use one of these:
* https://github.com/bitshares/bitshares-core/blob/2.0.181127/libraries/db/include/graphene/db/object_id.hpp#L83
* https://github.com/bitshares/bitshares-core/blob/2.0.181127/libraries/db/include/graphene/db/object_id.hpp#L192

File: libraries/chain/htlc_evaluator.cpp [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      void_result htlc_redeem_evaluator::do_evaluate(const htlc_redeem_operation& o)
+      {
+    	   htlc_obj = &db().get<htlc_object>(o.htlc_id);
+
+  		   FC_ASSERT(o.preimage.size() == htlc_obj->preimage_size, "Preimage size mismatch.");
+    		FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "Preimage provided after escrow expiration.");
+
+    		// see if the preimages match
+         bool match = false;
+         if (htlc_obj->preimage_hash_algorithm == graphene::chain::hash_algorithm::sha256)
```
Comment (pmconrad) [inline libraries/chain/htlc_evaluator.cpp]: Use `switch` instead of `if` cascade?

File: libraries/chain/htlc_evaluator.cpp [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+  		   FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "HTLC has already expired.");
+
+    	   return void_result();
+      }
+
+      void_result htlc_extend_evaluator::do_apply(const htlc_extend_operation& o)
+      {
+         db().modify(*htlc_obj, [&o](htlc_object& db_obj) {
+            db_obj.expiration = fc::time_point_sec(db_obj.expiration.sec_since_epoch() + o.seconds_to_add);
```
Comment (pmconrad) [inline libraries/chain/htlc_evaluator.cpp]: Can add seconds directly to `fc::time_point_sec`: https://github.com/bitshares/bitshares-fc/blob/master/include/fc/time.hpp#L103

File: tests/tests/htlc_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+   {
+
+      graphene::chain::committee_member_update_global_parameters_operation param_op;
+
+      graphene::chain::proposal_create_operation create_operation;
+
+      create_operation.fee_paying_account = committee_account;
+      create_operation.review_period_seconds = 60 * 60 * 48;
+      create_operation.proposed_ops.emplace_back(param_op);
+   }
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: Add check for other htlc operations inside proposal. And for update_parameters with fees for new operations.

File: tests/tests/htlc_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+      create_operation.key_hash = key_hash;
+      create_operation.hash_type = graphene::chain::hash_algorithm::sha256;
+      create_operation.key_size = key_size;
+      create_operation.source = alice_id;
+      trx.operations.push_back(create_operation);
+      sign(trx, alice_private_key);
+      PUSH_TX(db, trx, ~0);
+      trx.clear();
+      graphene::chain::signed_block blk = generate_block();
+      // can we assume that alice's transaction will be the only one in this block?
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: Well, it's your testcase. :-)

File: tests/tests/htlc_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+
+   // make sure Bob (or anyone) can see the details of the transaction
+   graphene::app::database_api db_api(db);
+   optional<graphene::chain::htlc_object> htlc = db_api.get_htlc(fc::json::to_pretty_string(alice_htlc_id));
+   BOOST_CHECK(htlc);
+
+   // let it expire (wait for timeout)
+   generate_blocks(fc::time_point_sec(120) );
+   // verify funds return (minus the fees)
+   BOOST_CHECK_EQUAL( get_balance(alice_id, graphene::chain::asset_id_type()), 90000 );
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: Looks like nothing was returned, probably because `generate_blocks( fc::time_point_sec(120) )` didn't generate enough. Hint: try `db.head_block_time + 120`

File: tests/tests/htlc_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+      update_operation.htlc_id = alice_htlc_id;
+      update_operation.preimage = pre_image;
+      update_operation.fee = db.current_fee_schedule().calculate_fee( update_operation );
+      trx.operations.push_back(update_operation);
+      sign(trx, bob_private_key);
+      try
+      {
+         PUSH_TX(db, trx, ~0);
+      } 
+      catch (fc::exception& ex)
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: Better wrap the entire test case in `try { ... } FC_LOG_AND_RETHROW` or something.

File: tests/tests/htlc_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+
+      create_operation.amount = graphene::chain::asset( 100000 );
+      create_operation.destination = bob_id;
+      create_operation.seconds_in_force = 86400;
+      create_operation.key_hash = key_hash;
+      create_operation.hash_type = graphene::chain::hash_algorithm::sha256;
+      create_operation.key_size = key_size;
+      create_operation.source = alice_id;
+      create_operation.fee = db.current_fee_schedule().calculate_fee( create_operation );
+      trx.operations.push_back(create_operation);
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: Please BOOST_CHECK the actual fee on the operation, otherwise it's hard to follow the numbers.

File: libraries/chain/htlc_evaluator.cpp [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+   namespace chain {
+
+      optional<htlc_options> get_committee_htlc_options(graphene::chain::database& db)
+      {
+         return db.get_global_properties().parameters.extensions.value.updatable_htlc_options;
+      }
+
+      void_result htlc_create_evaluator::do_evaluate(const htlc_create_operation& o)
+      {
+         //FC_ASSERT( db().head_block_time() > HARDFORK_HTLC_TIME,
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/htlc_evaluator.cpp]: I did this for testing. Without the committee setting their parameters, they won't get much further. The committee cannot set the parameters until (a) hardfork time or (b) modifying the genesis block (how we do it in our current private testnet). I am fine if you want to have this protection here, but I am considering removing it. I will defer to you.

File: libraries/chain/htlc_evaluator.cpp [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+         }
+         return true;
+      }
+
+      void_result htlc_redeem_evaluator::do_evaluate(const htlc_redeem_operation& o)
+      {
+    	   htlc_obj = &db().get<htlc_object>(o.htlc_id);
+
+  		   FC_ASSERT(o.preimage.size() == htlc_obj->preimage_size, "Preimage size mismatch.");
+    		FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "Preimage provided after escrow expiration.");
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/htlc_evaluator.cpp]: That depends on when head_block_time is updated. Is it after all transactions in the block have been processed? If so, I agree that this would be redundant. If you don't know off the top of your head, I'll research.

File: libraries/chain/include/graphene/chain/protocol/fee_schedule.hpp [hunk 15]
```diff
@@ -43,6 +43,12 @@ namespace graphene { namespace chain {
          FC_ASSERT( itr != parameters.end() );
          return itr->template get<typename Operation::fee_parameters_type>();
       }
+      typename Operation::fee_parameters_type& get(flat_set<fee_parameters>& parameters)
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/fee_schedule.hpp]: Made const. Were you also asking why the get is here?

File: libraries/chain/db_notify.cpp [hunk 16]
```diff
@@ -256,6 +256,18 @@ struct get_impacted_account_visitor
    {
       _impacted.insert( op.fee_payer() ); // account_id
    }
+   void operator()( const htlc_create_operation& op )
+   {
+      _impacted.insert( op.fee_payer() ); 
+   }
+   void operator()( const htlc_redeem_operation& op )
+   {
+      _impacted.insert( op.fee_payer() );
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/db_notify.cpp]: How would this operation get access to the database to look up the creator? Do you know of an example I could build off of?

File: libraries/chain/include/graphene/chain/htlc_object.hpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+} } // namespace graphene::chain
+
+namespace fc
+{
+  template<> 
+  struct get_typename<fc::enum_type<unsigned char, graphene::chain::hash_algorithm>> 
+  { 
+     static const char* name()
+     { 
+        return "fc::enum_type<unsigned char, graphene::chain::hash_algorithm>";
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/htlc_object.hpp]: I believe we can name it whatever we want. But if we ever want to debug serialization, I think it may help to use the official "what it was". I have not done a deep dive into un/serialization, so not sure of all places it is used, but I do know than it needs to be here and return something.

File: libraries/wallet/wallet.cpp (hunk 4 above)
Comment (jmjatlanta) [reply libraries/wallet/wallet.cpp]: Well, almost. It uses char instead of uint8_t. I modified string_to_vec to use fc::from_hex. I will revisit this to see if there is a better way.

File: libraries/chain/protocol/htlc.cpp [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+   void htlc_create_operation::validate()const {
+     FC_ASSERT( source != destination );
+      FC_ASSERT( fee.amount >= 0 );
+      FC_ASSERT( amount.amount > 0 );
+   }
+
+   void htlc_redeem_operation::validate()const {
+      FC_ASSERT( fee.amount >= 0 );
+      FC_ASSERT( preimage.size() > 0 );
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/protocol/htlc.cpp]: Yuck. But yes, you are right. Fixed.

File: libraries/chain/htlc_evaluator.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/htlc_evaluator.cpp]: Agree to remove.

File: libraries/chain/include/graphene/chain/htlc_object.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/htlc_object.hpp]: AFAIK the JS library uses the output from `js_operation_serializer`, so we must provide something that's meaningful to them (and matches the serialized form).

File: libraries/chain/include/graphene/chain/protocol/fee_schedule.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/fee_schedule.hpp]: Yes. There is already cget for const access. If you need get it means you're accessing it as non-const somewhere, which is probably an error.

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+          * @brief who will pay the fee
+          */
+         account_id_type fee_payer()const { return source; }
+
+         /****
+          * @brief calculates the fee to be paid for this operation
+          */
+         share_type calculate_fee(const fee_parameters_type& fee_params)const
+         {
+            uint32_t days = seconds_in_force / (60 * 60 * 24);
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: And watch out for overflows when multiplying first.

File: libraries/chain/db_notify.cpp [hunk 20]
```diff
@@ -256,6 +256,22 @@ struct get_impacted_account_visitor
...
+   {
+      _impacted.insert( op.fee_payer() ); 
+   }
+   void operator()( const htlc_extend_operation& op )
+   {
+      _impacted.insert( op.fee_payer() ); 
+   }
+   void operator()( const htlc_refund_operation& op ) 
+   { 
+      _impacted.insert( op.fee_payer() ); // bidder
```
Comment (pmconrad) [inline libraries/chain/db_notify.cpp]: Who is "bidder"?

File: libraries/chain/htlc_evaluator.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/htlc_evaluator.cpp]: It's updated in `update_global_dynamic_data`, i. e. after tx processing but before `clear_expired_*`.

File: libraries/chain/htlc_evaluator.cpp [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+         return void_result();
+      }
+
+      object_id_type htlc_create_evaluator::do_apply(const htlc_create_operation& o)
+      {
+         try {
+            graphene::chain::database& dbase = db();
+            dbase.adjust_balance( o.from, -o.amount );
+
+            const htlc_object& esc = db().create<htlc_object>([&dbase,o]( htlc_object& esc ) {
```
Comment (pmconrad) [inline libraries/chain/htlc_evaluator.cpp]: `o` should be captured by reference, too.

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 22]
```diff
@@ -73,7 +73,7 @@ namespace graphene {
           */
          share_type calculate_fee(const fee_parameters_type& fee_params)const
          {
-            uint32_t days = std::min<uint64_t>(1, claim_period_seconds / (60 * 60 * 24));
+            uint64_t days = std::max<uint64_t>(1, claim_period_seconds / (60 * 60 * 24));
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: Hm, perhaps always rounding up to the next full day might be acceptable too.

File: libraries/chain/db_notify.cpp [hunk 23]
```diff
@@ -344,6 +360,12 @@ void get_relevant_accounts( const object* obj, flat_set<account_id_type>& accoun
         } case balance_object_type:{
            /** these are free from any accounts */
            break;
+        } case htlc_object_type: {
+              const auto& htlc_obj = dynamic_cast<const htlc_object*>(obj);
+              FC_ASSERT( htlc_obj != nullptr);
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/db_notify.cpp]: Arguing with myself as to make it look like all the others or pick a standard and any new code follows the standard. I decided to make it look like the others in the case statement (recently fixed the space before the curly brace. Should I make new code look like the old when in the same method, or always stick to the standard?

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <fc/time.hpp>
+#include <boost/container/flat_set.hpp>
+#include <graphene/chain/htlc_object.hpp>
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: renamed htlc_hash_algorithm

File: libraries/chain/db_notify.cpp (hunk 20 above)
Comment (jmjatlanta) [reply libraries/chain/db_notify.cpp]: I have no idea where that came from. But it no longer exists.

File: tests/tests/htlc_tests.cpp [hunk 25]
```diff
@@ -0,0 +1,457 @@
...
+std::vector<uint8_t> hash_it(std::vector<uint8_t> preimage)
+{
+	fc::sha256 hash = fc::sha256::hash((char*)preimage.data(), preimage.size());
+   return std::vector<uint8_t>(hash.data(), hash.data() + hash.data_size());
+}
+
+void set_committee_parameters(database_fixture* db_fixture)
+{
+   // set the committee parameters
+   db_fixture->db.modify(db_fixture->db.get_global_properties(), [](global_property_object& p) {
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply tests/tests/htlc_tests.cpp]: Done

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 26]
```diff
@@ -112,8 +120,9 @@ namespace graphene {
          share_type calculate_fee(const fee_parameters_type& fee_params)const
          {
             if (fee_params.fee_per_kb > 0)
-               return std::max<share_type>(fee_params.fee_per_kb, preimage.size() / 1024 * fee_params.fee_per_kb);
-            return 0;
+               return fee_params.fee + std::max<share_type>(fee_params.fee_per_kb, 
+                     std::ceil( preimage.size() / (double)1024 ) * fee_params.fee_per_kb );
```
Comment (pmconrad) [inline libraries/chain/include/graphene/chain/protocol/htlc.hpp]: No floating point please. `(preimage.size() + 1023) / 1024` achieves the same.

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+          */
+         account_id_type fee_payer()const { return update_issuer; }
+
+         /****
+          * @brief calculates the fee to be paid for this operation
+          */
+         share_type calculate_fee(const fee_parameters_type& fee_params)const
+         {
+            uint32_t days = seconds_to_add / (60 * 60 * 24);
+            return fee_params.fee + (fee_params.fee_per_day * days);
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: When rounding up, `max` is not needed anymore. Also, no floating point please.

**Comment by pmconrad** [comment]
Missing a test case for `htlc_extend`

File: libraries/chain/htlc_evaluator.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/htlc_evaluator.cpp]: Removed redundant check

**Comment by jmjatlanta** [comment]
> Missing a test case for `htlc_extend`

added to htlc_tests/htlc_fulfilled

File: tests/tests/htlc_tests.cpp (hunk 8 above)
Comment (jmjatlanta) [reply tests/tests/htlc_tests.cpp]: Merged in your changes that include htlc proposal tests

File: libraries/chain/db_notify.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply libraries/chain/db_notify.cpp]: added virtual op htlc_redeemed_operation

File: tests/tests/htlc_tests.cpp [hunk 28]
```diff
@@ -263,32 +264,49 @@ BOOST_AUTO_TEST_CASE( htlc_fulfilled )
...
       try
       {
-         PUSH_TX(db, trx, ~0);
+         PUSH_TX( db, trx, ~0 );
       } 
-      catch (fc::exception& ex)
+      catch ( fc::exception& ex )
       {
-         BOOST_FAIL(ex.what());
+         BOOST_FAIL( ex.to_detail_string() );
```
Comment (pmconrad) [inline tests/tests/htlc_tests.cpp]: I think this try/catch/FAIL is superfluous, same effect should already be achieved by `FC_LOG_AND_RETHROW`

File: libraries/chain/protocol/htlc.cpp [hunk 29]
```diff
@@ -28,21 +28,21 @@
 namespace graphene { namespace chain {
 
    void htlc_create_operation::validate()const {
-      FC_ASSERT( fee.amount >= 0 );
-      FC_ASSERT( amount.amount > 0 );
+      FC_ASSERT( fee.amount >= 0, "Fee amount should not be zero" );
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: "...should not be negative"

File: libraries/chain/protocol/htlc.cpp [hunk 30]
```diff
@@ -28,21 +28,21 @@
...
+      FC_ASSERT( amount.amount > 0, "HTLC amount should not be zero" );
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: "...should be positive|greater than zero"

File: libraries/chain/protocol/htlc.cpp [hunk 31]
```diff
@@ -28,21 +28,21 @@
...
       // multiply with overflow check
       uint64_t per_day_fee = fee_params.fee_per_day * days;
-      FC_ASSERT( days == 0 || per_day_fee / fee_params.fee_per_day == days );
+      FC_ASSERT( days == 0 || per_day_fee / days == fee_params.fee_per_day, "Fee calculation overflow" );
       return fee_params.fee + per_day_fee;
    }
 
    void htlc_redeem_operation::validate()const {
-      FC_ASSERT( fee.amount >= 0 );
+      FC_ASSERT( fee.amount >= 0, "Fee amount should not be zero" );
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: "...should not be negative"

File: libraries/chain/protocol/htlc.cpp [hunk 32]
```diff
@@ -52,14 +52,14 @@ namespace graphene { namespace chain {
    }
 
    void htlc_extend_operation::validate()const {
-      FC_ASSERT( fee.amount >= 0 );
+      FC_ASSERT( fee.amount >= 0 , "Fee amount should not be zero");
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: "...should not be negative"

File: libraries/chain/protocol/htlc.cpp [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+   }
+
+   void htlc_redeem_operation::validate()const {
+      FC_ASSERT( fee.amount >= 0, "Fee amount should not be zero" );
+   }
+
+   share_type htlc_redeem_operation::calculate_fee( const fee_parameters_type& fee_params )const
+   {
+      return fee_params.fee
+             + ( preimage.size() + 1023 ) / 1024 * fee_params.fee_per_kb;
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: Please add similiar overflow check here too

File: tests/tests/htlc_tests.cpp (hunk 28 above)
Comment (jmjatlanta) [reply tests/tests/htlc_tests.cpp]: Removed internal try/catch

File: libraries/chain/protocol/htlc.cpp (hunk 33 above)
Comment (jmjatlanta) [reply libraries/chain/protocol/htlc.cpp]: Done

File: libraries/chain/protocol/htlc.cpp [hunk 34]
```diff
@@ -42,17 +42,22 @@ namespace graphene { namespace chain {
...
-      FC_ASSERT( fee.amount >= 0, "Fee amount should not be zero" );
+      FC_ASSERT( fee.amount >= 0, "Fee amount should not be negative" );
    }
 
    share_type htlc_redeem_operation::calculate_fee( const fee_parameters_type& fee_params )const
    {
+      if (fee_params.fee_per_kb == 0)
+         return fee_params.fee;
+      uint64_t product = 1024 * fee_params.fee_per_kb;
+      FC_ASSERT( product / 1024 == fee_params.fee_per_kb, "Fee calculation overflow");
```
Comment (pmconrad) [inline libraries/chain/protocol/htlc.cpp]: Ah, no -
```
uint64_t kb = ( preimage.size() + 1023 ) / 1024;
uint64_t product = kb * fee_params.fee_per_kb;
FC_ASSERT( kb == 0 || product / kb == fee_params.fee_per_kb, "Fee calculation overflow");
```

**Comment by pmconrad** [comment]
Thanks!
